doc/generated/pmem_flush.3
.\" Automatically generated by Pandoc 2.0.6
.\"
.TH "PMEM_FLUSH" "3" "2020-01-16" "PMDK - pmem API version 1.1" "PMDK Programmer's Manual"
.hy
.\" Copyright 2014-2020, Intel Corporation
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in
.\"       the documentation and/or other materials provided with the
.\"       distribution.
.\"
.\"     * Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived
.\"       from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.SH NAME
.PP
\f[B]pmem_flush\f[](), \f[B]pmem_drain\f[](), \f[B]pmem_persist\f[](),
\f[B]pmem_msync\f[](), \f[B]pmem_deep_flush\f[](),
\f[B]pmem_deep_drain\f[](), \f[B]pmem_deep_persist\f[](),
\f[B]pmem_has_hw_drain\f[](), \f[B]pmem_has_auto_flush\f[]() \- check
persistency, store persistent data and delete mappings
.SH SYNOPSIS
.IP
.nf
\f[C]
#include\ <libpmem.h>

void\ pmem_persist(const\ void\ *addr,\ size_t\ len);
int\ pmem_msync(const\ void\ *addr,\ size_t\ len);
void\ pmem_flush(const\ void\ *addr,\ size_t\ len);
void\ pmem_deep_flush(const\ void\ *addr,\ size_t\ len);\ (EXPERIMENTAL)
int\ pmem_deep_drain(const\ void\ *addr,\ size_t\ len);\ (EXPERIMENTAL)
int\ pmem_deep_persist(const\ void\ *addr,\ size_t\ len);\ (EXPERIMENTAL)
void\ pmem_drain(void);
int\ pmem_has_auto_flush(void);\ (EXPERIMENTAL)
int\ pmem_has_hw_drain(void);
\f[]
.fi
.SH DESCRIPTION
.PP
The functions in this section provide access to the stages of flushing
to persistence, for the less common cases where an application needs
more control of the flushing operations than the \f[B]pmem_persist\f[]()
function.
.RS
.PP
WARNING: Using \f[B]pmem_persist\f[]() on a range where
\f[B]pmem_is_pmem\f[](3) returns false may not do anything useful \[en]
use \f[B]msync\f[](2) instead.
.RE
.PP
The \f[B]pmem_persist\f[]() function force any changes in the range
[\f[I]addr\f[], \f[I]addr\f[]+\f[I]len\f[]) to be stored durably in
persistent memory.
This is equivalent to calling \f[B]msync\f[](2) but may be more optimal
and will avoid calling into the kernel if possible.
There are no alignment restrictions on the range described by
\f[I]addr\f[] and \f[I]len\f[], but \f[B]pmem_persist\f[]() may expand
the range as necessary to meet platform alignment requirements.
.RS
.PP
WARNING: Like \f[B]msync\f[](2), there is nothing atomic or
transactional about this call.
Any unwritten stores in the given range will be written, but some stores
may have already been written by virtue of normal cache
eviction/replacement policies.
Correctly written code must not depend on stores waiting until
\f[B]pmem_persist\f[]() is called to become persistent \[en] they can
become persistent at any time before \f[B]pmem_persist\f[]() is called.
.RE
.PP
The \f[B]pmem_msync\f[]() function is like \f[B]pmem_persist\f[]() in
that it forces any changes in the range [\f[I]addr\f[],
\f[I]addr\f[]+\f[I]len\f[]) to be stored durably.
Since it calls \f[B]msync\f[](), this function works on either
persistent memory or a memory mapped file on traditional storage.
\f[B]pmem_msync\f[]() takes steps to ensure the alignment of addresses
and lengths passed to \f[B]msync\f[]() meet the requirements of that
system call.
It calls \f[B]msync\f[]() with the \f[B]MS_SYNC\f[] flag as described in
\f[B]msync\f[](2).
Typically the application only checks for the existence of persistent
memory once, and then uses that result throughout the program, for
example:
.IP
.nf
\f[C]
/*\ do\ this\ call\ once,\ after\ the\ pmem\ is\ memory\ mapped\ */
int\ is_pmem\ =\ pmem_is_pmem(rangeaddr,\ rangelen);

/*\ ...\ make\ changes\ to\ a\ range\ of\ pmem\ ...\ */

/*\ make\ the\ changes\ durable\ */
if\ (is_pmem)
\ \ \ \ pmem_persist(subrangeaddr,\ subrangelen);
else
\ \ \ \ pmem_msync(subrangeaddr,\ subrangelen);

/*\ ...\ */
\f[]
.fi
.RS
.PP
WARNING: On Linux, \f[B]pmem_msync\f[]() and \f[B]msync\f[](2) have no
effect on memory ranges mapped from Device DAX.
In case of memory ranges where \f[B]pmem_is_pmem\f[](3) returns true use
\f[B]pmem_persist\f[]() to force the changes to be stored durably in
persistent memory.
.RE
.PP
The \f[B]pmem_flush\f[]() and \f[B]pmem_drain\f[]() functions provide
partial versions of the \f[B]pmem_persist\f[]() function.
\f[B]pmem_persist\f[]() can be thought of as this:
.IP
.nf
\f[C]
void
pmem_persist(const\ void\ *addr,\ size_t\ len)
{
\ \ \ \ /*\ flush\ the\ processor\ caches\ */
\ \ \ \ pmem_flush(addr,\ len);

\ \ \ \ /*\ wait\ for\ any\ pmem\ stores\ to\ drain\ from\ HW\ buffers\ */
\ \ \ \ pmem_drain();
}
\f[]
.fi
.PP
These functions allow advanced programs to create their own variations
of \f[B]pmem_persist\f[]().
For example, a program that needs to flush several discontiguous ranges
can call \f[B]pmem_flush\f[]() for each range and then follow up by
calling \f[B]pmem_drain\f[]() once.
.PP
The semantics of \f[B]pmem_deep_flush\f[]() function is the same as
\f[B]pmem_flush\f[]() function except that \f[B]pmem_deep_flush\f[]() is
indifferent to \f[B]PMEM_NO_FLUSH\f[] environment variable (see
\f[B]ENVIRONMENT\f[] section in \f[B]libpmem\f[](7)) and always flushes
processor caches.
.PP
The behavior of \f[B]pmem_deep_persist\f[]() function is the same as
\f[B]pmem_persist\f[](), except that it provides higher reliability by
flushing persistent memory stores to the most reliable persistence
domain available to software rather than depending on automatic WPQ
(write pending queue) flush on power failure (ADR).
.PP
The \f[B]pmem_deep_flush\f[]() and \f[B]pmem_deep_drain\f[]() functions
provide partial versions of \f[B]pmem_deep_persist\f[]() function.
\f[B]pmem_deep_persist\f[]() can be thought of as this:
.IP
.nf
\f[C]
int\ pmem_deep_persist(const\ void\ *addr,\ size_t\ len)
{
\ \ \ \ /*\ flush\ the\ processor\ caches\ */
\ \ \ \ pmem_deep_flush(addr,\ len);

\ \ \ \ /*\ wait\ for\ any\ pmem\ stores\ to\ drain\ from\ HW\ buffers\ */
\ \ \ \ return\ pmem_deep_drain(addr,\ len);
}
\f[]
.fi
.PP
Since this operation is usually much more expensive than
\f[B]pmem_persist\f[](), it should be used rarely.
Typically the application should use this function only to flush the
most critical data, which are required to recover after the power
failure.
.PP
The \f[B]pmem_has_auto_flush\f[]() function checks if the machine
supports automatic CPU cache flush on power failure or system crash.
Function returns true only when each NVDIMM in the system is covered by
this mechanism.
.PP
The \f[B]pmem_has_hw_drain\f[]() function checks if the machine supports
an explicit \f[I]hardware drain\f[] instruction for persistent memory.
.SH RETURN VALUE
.PP
The \f[B]pmem_persist\f[]() function returns no value.
.PP
The \f[B]pmem_msync\f[]() return value is the return value of
\f[B]msync\f[](), which can return \-1 and set \f[I]errno\f[] to
indicate an error.
.PP
The \f[B]pmem_flush\f[](), \f[B]pmem_drain\f[]() and
\f[B]pmem_deep_flush\f[]() functions return no value.
.PP
The \f[B]pmem_deep_persist\f[]() and \f[B]pmem_deep_drain\f[]() return 0
on success.
Otherwise it returns \-1 and sets \f[I]errno\f[] appropriately.
If \f[I]len\f[] is equal zero \f[B]pmem_deep_persist\f[]() and
\f[B]pmem_deep_drain\f[]() return 0 but no flushing take place.
.PP
The \f[B]pmem_has_auto_flush\f[]() function returns 1 if given platform
supports processor cache flushing on a power loss event.
Otherwise it returns 0.
On error it returns \-1 and sets \f[I]errno\f[] appropriately.
.PP
The \f[B]pmem_has_hw_drain\f[]() function returns true if the machine
supports an explicit \f[I]hardware drain\f[] instruction for persistent
memory.
On Intel processors with persistent memory, stores to persistent memory
are considered persistent once they are flushed from the CPU caches, so
this function always returns false.
Despite that, programs using \f[B]pmem_flush\f[]() to flush ranges of
memory should still follow up by calling \f[B]pmem_drain\f[]() once to
ensure the flushes are complete.
As mentioned above, \f[B]pmem_persist\f[]() handles calling both
\f[B]pmem_flush\f[]() and \f[B]pmem_drain\f[]().
.SH SEE ALSO
.PP
\f[B]msync\f[](2), \f[B]pmem_is_pmem\f[](3), \f[B]libpmem\f[](7) and
\f[B]<http://pmem.io>\f[]


ChangeLog
Fri Jan 17 2020 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.8-rc1

	This release:
	- Introduces new API in libpmemobj for user-assisted defragmentation
	  (see pmemobj_defrag man page for details).
	- Introduces experimental support for PPC64LE architecture.
	- Introduces new flag in libpmemobj for opting-out of transaction aborts
	  on failure (POBJ_TX_NO_ABORT), along with new variants of existing APIs
	  that didn't accept flags (pmemobj_tx_xfree, pmemobj_tx_xpublish,
	  pmemobj_tx_xlock, pmemobj_tx_xlog_append_buffer, pmemobj_tx_xstrdup,
	  pmemobj_tx_xwcsdup).
	- Moves out libvmem and libvmmalloc to the new repository
	  (https://github.com/pmem/vmem).

	Other changes:
	- obj: introduce new statistics useful for defrag
	- obj: introduce transient/persistent statistics enabling
	- obj: introduce pmemobj_tx_(set/get)_user_data funcs
	- obj: introduce pmemobj_(set/get)_user_data funcs
	- obj: disable workaround for offsetof() since VS 15.5 in public header
	- common: drop support for libndctl < 63 on Linux
	- pool: rename -N --no-exec to -d --dry-run

	Notable bug fixes:
	- obj: fix potential NULL-dereference in ulog_store
	- obj: fix unintended recursive locking during palloc
	- obj: fix lock release order in palloc publish
	- obj: fix transient redo log of more than 64 entries
	- obj: fix capacity ulog calculation
	- obj: fix check of unaligned capacity size
	- rpmem: add a missing case for GPSPM + FLUSH_STRICT
	- pmem: fix pmemcheck support on platforms with eADR
	- pool: fix possible memory leak
	- rpmem: fix possible memory leak in rpmemd_config_read
	- rpmem: fix possible memory leak in rpmemd_log_init
	- rpmem: fix possible use-after-free

Mon Sep 30 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.7

	This release:
	- Introduces new APIs in libpmemobj for managing space used by transactions.
	  (see pmemobj_tx_log_append_buffer man page for details)
	- Introduces new APIs in librpmem, splitting rpmem_persist into rpmem_flush
	  and rpmem_drain, allowing applications to use the flush + drain model
	  already known from libpmem. (libpmemobj does not use this feature yet)
	- Optimizes large libpmemobj transactions by significantly reducing
	  the amount of memory modified at the commit phase.
	- Optimizes tracking of libpmemobj reservations.
	- Adds new flags for libpmemobj's pmemobj_tx_xadd_range[_direct] API:
	  POBJ_XADD_NO_SNAPSHOT and POBJ_XADD_ASSUME_INITIALIZED, allowing
	  applications to optimize how memory is tracked by the library.

	To support some of the above changes the libpmemobj on-media layout had
	to be changed, which means that old pools have to be converted using
	pmdk-convert >= 1.7.

	Other changes:
	- obj: fix merging of ranges when NOFLUSH flag is used (#1100)
	- rpmem: fix closing of ssh connection (#995, #1060)
	- obj: abort transaction on pmemobj_tx_publish failure

	Internal changes:
	- test: fault injection tests for pmemblk, pmemlog, and pmemobj
	- test: improved Python testing framework
	- test: support real pmem in bad blocks tests
	- common: allow not building examples and benchmarks

Tue Aug 27 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.6.1

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains,
	incorporates build fixes for FreeBSD and fixes a number of
	smaller bugs.

	Detailed list of bug fixes:
	- common: (win) fix possible pool file coruption (#972, #715, #603)
	- common: implement correct / robust device_dax_alignment (#1071)
	- obj: fix recycler not locating unused chunks
	- doc: update pmemobj_tx_lock documentation wrt behavior on fail
	- common: fix persistent domain detection (#1093)
	- common: vecq: fix a pointer-to-struct aliasing violation
	  (crash on arm64)
	- common: fix minor issues related to ndctl linking
	- obj: drop recursion from pmemobj_next
	- common: fix bug in badblock file error handling
	- obj: fix handling of malloc failures
	- common: fix handling of malloc failures (ctl)
	- jemalloc: fix build with gcc 9
	- obj: don't overwrite errno when palloc_heap_check_remote fails
	- doc: fix pmreorder emit log macro
	- rpmem: change order of rpmem init (#1043)
	- common: Fix build failure due to unused macro PAGE_SIZE
	- common: support older versions of pkg-config
	- tools: link with release variant of pmemcommon
	- common: add PMDK prefix to local copy of queue.h (#990)
	- rpmem: switch to using an open coded basename (FreeBSD)
	- common: posix_fallocate: guard against integer underflow in
	  check (FreeBSD)
	- test: support Valgrind 3.15
	- test: skip if fi_info is missing
	- test: (win) fix sparsefile error handling
	- test: fix libpmempool_feature tests that match logs
	- test: remove vmem_delete test (#1074)
	- test: adjust matchfiles in vmem_valgrind_region test (#1087)
	- test: remove old log files for windows (#1013)
	- test: remove invalid expect_normal_exit (#1092)
	- test: suppress ld leak (#1098)
	- test: Expose necessary symbols in libvmmalloc_dummy_funcs
	  (FreeBSD)
	- test: fix tests failing because `tput` fails (FreeBSD)
	- test: avoid obj_critnib_mt taking very long on many-core machines
	- test: deal with libndctl's path without build system
	- test: overwrite old log in pmempool_create/TEST14.PS1
	- test: fix match files in tests which use dax devices
	- test: fix match file in rpmem_addr_ext test
	- test: fix pmempool_check test

Wed Aug 28 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.5.2

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains

src/benchmarks/pmem_flush.c
/*
 * Copyright 2016, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *	* Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *	* Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 *	* Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived
 *        from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * pmem_flush.c -- benchmark implementation for pmem_persist and pmem_msync
 */
#include <libpmem.h>
#include <fcntl.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <limits.h>
#include <assert.h>
#include <sys/mman.h>

#include "benchmark.h"

#define PAGE_4K ((uintptr_t)1 << 12)
#define PAGE_2M ((uintptr_t)1 << 21)

/*
 * align_addr -- round addr down to given boundary
 */
static char *
align_addr(char *addr, uintptr_t align)
{
	return (char *)((uintptr_t)addr & ~(align - 1));
}

/*
 * align_len -- increase len by the amount we gain when we round addr down
 */
static size_t
align_len(size_t len, char *addr, uintptr_t align)
{
	return len + ((uintptr_t)addr & (align - 1));
}

/*
 * roundup_len -- increase len by the amount we gain when we round addr down,
 *                then round up to the nearest multiple of 4K
 */
static size_t
roundup_len(size_t len, char *addr, uintptr_t align)
{
	return (align_len(len, addr, align) + align - 1) & ~(align - 1);
}

/*
 * pmem_args -- benchmark specific arguments
 */
struct pmem_args
{
	char *operation;	/* msync, dummy_msync, persist, ... */
	char *mode;		/* stat, seq, rand */
	bool no_warmup;		/* don't do warmup */
};

/*
 * pmem_bench -- benchmark context
 */
struct pmem_bench
{
	uint64_t *offsets;	/* write offsets */
	size_t n_offsets;	/* number of elements in offsets array */
	size_t fsize;		/* The size of the allocated PMEM */

	struct pmem_args *pargs;	/* prog_args structure */

	void *pmem_addr;	/* PMEM base address */
	size_t pmem_len;	/* length of PMEM mapping */

	void *invalid_addr;	/* invalid pages */
	void *nondirty_addr;	/* non-dirty pages */

	void *pmem_addr_aligned;	/* PMEM pages - 2M aligned */
	void *invalid_addr_aligned;	/* invalid pages - 2M aligned */
	void *nondirty_addr_aligned;	/* non-dirty pages - 2M aligned */

	/* the actual benchmark operation */
	int (*func_op) (struct pmem_bench *pmb, void *addr, size_t len);
};


/*
 * mode_seq -- if copy mode is sequential, returns index of a chunk.
 */
static uint64_t
mode_seq(struct pmem_bench *pmb, uint64_t index)
{
	return index;
}

/*
 * mode_stat -- if mode is static, the offset is always 0
 */
static uint64_t
mode_stat(struct pmem_bench *pmb, uint64_t index)
{
	return 0;
}

/*
 * mode_rand -- if mode is random, returns index of a random chunk
 */
static uint64_t
mode_rand(struct pmem_bench *pmb, uint64_t index)
{
	return rand() % pmb->n_offsets;
}

/*
 * operation_mode -- the mode of the copy process
 *
 *	* static     - write always the same chunk,
 *	* sequential - write chunk by chunk,
 *	* random     - write to chunks selected randomly.
 */
struct op_mode {
	const char *mode;
	uint64_t (*func_mode) (struct pmem_bench *pmb, uint64_t index);
};

static struct op_mode modes[] = {
	{ "stat", mode_stat },
	{ "seq", mode_seq },
	{ "rand", mode_rand },
};

#define MODES (sizeof(modes) / sizeof(modes[0]))

/*
 * parse_op_mode -- parses command line "--mode"
 * and returns proper operation mode index.
 */
static int
parse_op_mode(const char *arg)
{
	for (int i = 0; i < MODES; i++) {
		if (strcmp(arg, modes[i].mode) == 0)
			return i;
	}
	return -1;
}

/*
 * flush_noop -- dummy flush, does nothing
 */
static int
flush_noop(struct pmem_bench *pmb, void *addr, size_t len)
{
	return 0;
}

/*
 * flush_persist -- flush data to persistence using pmem_persist()
 */
static int
flush_persist(struct pmem_bench *pmb, void *addr, size_t len)
{
	pmem_persist(addr, len);
	return 0;
}

/*
 * flush_persist_4K -- always flush entire 4K page(s) using pmem_persist()
 */
static int
flush_persist_4K(struct pmem_bench *pmb, void *addr, size_t len)
{
	void *ptr = align_addr(addr, PAGE_4K);
	len = roundup_len(len, addr, PAGE_4K);

	pmem_persist(ptr, len);
	return 0;
}

/*
 * flush_persist_2M -- always flush entire 2M page(s) using pmem_persist()
 */
static int
flush_persist_2M(struct pmem_bench *pmb, void *addr, size_t len)
{
	void *ptr = align_addr(addr, PAGE_2M);
	len = roundup_len(len, addr, PAGE_2M);

	pmem_persist(ptr, len);
	return 0;
}

/*
 * flush_msync -- flush data to persistence using pmem_msync()
 */
static int
flush_msync(struct pmem_bench *pmb, void *addr, size_t len)
{
	pmem_msync(addr, len);
	return 0;
}

/*
 * flush_msync_async -- emulate dummy msync() using MS_ASYNC flag
 */
static int
flush_msync_async(struct pmem_bench *pmb, void *addr, size_t len)
{
	void *ptr = align_addr(addr, PAGE_4K);
	len = align_len(len, addr, PAGE_4K);

	msync(ptr, len, MS_ASYNC);
	return 0;
}

/*
 * flush_msync_0 -- emulate dummy msync() using zero length
 */
static int
flush_msync_0(struct pmem_bench *pmb, void *addr, size_t len)
{
	void *ptr = align_addr(addr, PAGE_4K);
	len = align_len(len, addr, PAGE_4K);

	msync(ptr, 0, MS_SYNC);
	return 0;
}

/*
 * flush_persist_4K_msync_0 -- emulate msync() that only flushes CPU cache
 *
 * Do flushing in user space (4K pages) + dummy syscall.
 */
static int
flush_persist_4K_msync_0(struct pmem_bench *pmb, void *addr, size_t len)
{
	void *ptr = align_addr(addr, PAGE_4K);
	len = roundup_len(len, addr, PAGE_4K);

	pmem_persist(ptr, len);
	msync(ptr, 0, MS_SYNC);
	return 0;
}

/*
 * flush_persist_2M_msync_0 -- emulate msync() that only flushes CPU cache
 *
 * Do flushing in user space (2M pages) + dummy syscall.
 */
static int
flush_persist_2M_msync_0(struct pmem_bench *pmb, void *addr, size_t len)
{
	void *ptr = align_addr(addr, PAGE_2M);
	len = roundup_len(len, addr, PAGE_2M);

	pmem_persist(ptr, len);
	msync(ptr, 0, MS_SYNC);
	return 0;
}

/*
 * flush_msync_err -- emulate dummy msync() using invalid flags
 */
static int
flush_msync_err(struct pmem_bench *pmb, void *addr, size_t len)
{
	void *ptr = align_addr(addr, PAGE_4K);
	len = align_len(len, addr, PAGE_4K);

	msync(ptr, len, MS_SYNC|MS_ASYNC);
	return 0;
}

/*
 * flush_msync_nodirty -- call msync() on non-dirty pages
 */
static int
flush_msync_nodirty(struct pmem_bench *pmb, void *addr, size_t len)
{
	uintptr_t uptr = (uintptr_t)addr - (uintptr_t)pmb->pmem_addr_aligned;
	uptr += (uintptr_t)pmb->nondirty_addr_aligned;

	void *ptr = align_addr((void *)uptr, PAGE_4K);
	len = align_len(len, (void *)uptr, PAGE_4K);

	pmem_msync(ptr, len);
	return 0;
}

/*
 * flush_msync_invalid -- emulate dummy msync() using invalid address
 */
static int
flush_msync_invalid(struct pmem_bench *pmb, void *addr, size_t len)
{
	uintptr_t uptr = (uintptr_t)addr - (uintptr_t)pmb->pmem_addr_aligned;
	uptr += (uintptr_t)pmb->invalid_addr_aligned;

	void *ptr = align_addr((void *)uptr, PAGE_4K);
	len = align_len(len, (void *)uptr, PAGE_4K);

	pmem_msync(ptr, len);
	return 0;
}

struct op {
	const char *opname;
	int (*func_op) (struct pmem_bench *pmb, void *addr, size_t len);
};

static struct op ops[] = {
	{ "noop", flush_noop },
	{ "persist", flush_persist },
	{ "persist_4K", flush_persist_4K },
	{ "persist_2M", flush_persist_2M },
	{ "msync", flush_msync },
	{ "msync_0", flush_msync_0 },
	{ "msync_err", flush_msync_err },
	{ "persist_4K_msync_0", flush_persist_4K_msync_0 },
	{ "persist_2M_msync_0", flush_persist_2M_msync_0 },
	{ "msync_async", flush_msync_async },
	{ "msync_nodirty", flush_msync_nodirty },
	{ "msync_invalid", flush_msync_invalid },
};

#define NOPS (sizeof(ops) / sizeof(ops[0]))

/*
 * parse_op_type -- parses command line "--operation" argument
 * and returns proper operation type.
 */
static int
parse_op_type(const char *arg)
{
	for (int i = 0; i < NOPS; i++) {
		if (strcmp(arg, ops[i].opname) == 0)
			return i;
	}
	return -1;
}

/*
 * pmem_flush_init -- benchmark initialization
 *
 * Parses command line arguments, allocates persistent memory, and maps it.
 */
static int
pmem_flush_init(struct benchmark *bench, struct benchmark_args *args)
{
	assert(bench != NULL);
	assert(args != NULL);

	uint64_t (*func_mode) (struct pmem_bench *pmb, uint64_t index);

	struct pmem_bench *pmb = malloc(sizeof(struct pmem_bench));
	assert(pmb != NULL);

	pmb->pargs = args->opts;
	assert(pmb->pargs != NULL);

	int i = parse_op_type(pmb->pargs->operation);
	if (i == -1) {
		fprintf(stderr, "wrong operation: %s\n", pmb->pargs->operation);
		goto err_free_pmb;
	}
	pmb->func_op = ops[i].func_op;

	pmb->n_offsets = args->n_ops_per_thread * args->n_threads;

	pmb->fsize = pmb->n_offsets * args->dsize + (2 * PAGE_2M);

	/* round up to 2M boundary */
	pmb->fsize = (pmb->fsize + PAGE_2M - 1) & ~(PAGE_2M - 1);

	i = parse_op_mode(pmb->pargs->mode);
	if (i == -1) {
		fprintf(stderr, "wrong mode: %s\n", pmb->pargs->mode);
		goto err_free_pmb;
	}
	func_mode = modes[i].func_mode;

	/* populate offsets array */
	pmb->offsets = malloc(pmb->n_offsets * sizeof(*pmb->offsets));
	assert(pmb->offsets != NULL);

	for (size_t i = 0; i < pmb->n_offsets; ++i)
		pmb->offsets[i] = func_mode(pmb, i);

	/* create a pmem file and memory map it */
	pmb->pmem_addr = pmem_map_file(args->fname, pmb->fsize,
			PMEM_FILE_CREATE|PMEM_FILE_EXCL, args->fmode,
			&pmb->pmem_len, NULL);

	if (pmb->pmem_addr == NULL) {
		perror("pmem_map_file");
		goto err_free_pmb;
	}

	pmb->nondirty_addr = mmap(NULL, pmb->fsize, PROT_READ|PROT_WRITE,
			MAP_PRIVATE|MAP_ANON, -1, 0);

	if (pmb->nondirty_addr == MAP_FAILED) {
		perror("pmem_map1");
		goto err_unmap1;
	}

	pmb->invalid_addr = mmap(NULL, pmb->fsize, PROT_READ|PROT_WRITE,
			MAP_PRIVATE|MAP_ANON, -1, 0);

	if (pmb->invalid_addr == MAP_FAILED) {
		perror("pmem_map2");
		goto err_unmap2;
	}
	munmap(pmb->invalid_addr, pmb->fsize);

	pmb->pmem_addr_aligned =
		(void *)(((uintptr_t)pmb->pmem_addr + PAGE_2M - 1)
		& ~(PAGE_2M - 1));

	pmb->nondirty_addr_aligned =
		(void *)(((uintptr_t)pmb->nondirty_addr + PAGE_2M - 1)
		& ~(PAGE_2M - 1));

	pmb->invalid_addr_aligned =
		(void *)(((uintptr_t)pmb->invalid_addr + PAGE_2M - 1)
		& ~(PAGE_2M - 1));

	pmembench_set_priv(bench, pmb);

	if (!pmb->pargs->no_warmup) {
		size_t off;
		for (off = 0; off < pmb->fsize - PAGE_2M; off += PAGE_4K) {
			*(int *)((char *)pmb->pmem_addr_aligned + off) = 0;
			*(int *)((char *)pmb->nondirty_addr_aligned + off) = 0;
		}
	}

	return 0;

err_unmap2:
	munmap(pmb->nondirty_addr, pmb->fsize);
err_unmap1:
	pmem_unmap(pmb->pmem_addr, pmb->pmem_len);
err_free_pmb:
	free(pmb);

	return -1;
}

/*
 * pmem_flush_exit -- benchmark cleanup
 */
static int
pmem_flush_exit(struct benchmark *bench, struct benchmark_args *args)
{
	struct pmem_bench *pmb = (struct pmem_bench *)pmembench_get_priv(bench);
	pmem_unmap(pmb->pmem_addr, pmb->pmem_len);
	munmap(pmb->nondirty_addr, pmb->pmem_len);
	free(pmb);
	return 0;
}

/*
 * pmem_flush_operation -- actual benchmark operation
 */
static int
pmem_flush_operation(struct benchmark *bench, struct operation_info *info)
{
	struct pmem_bench *pmb = (struct pmem_bench *)pmembench_get_priv(bench);

	int op_idx = info->index;
	assert(op_idx < pmb->n_offsets);

	uint64_t chunk_idx = pmb->offsets[op_idx];
	void *addr = (char *)pmb->pmem_addr_aligned
					+ chunk_idx * info->args->dsize;

	/* store + flush */
	*(int *)addr = *(int *)addr + 1;
	pmb->func_op(pmb, addr, info->args->dsize);
	return 0;
}

/* structure to define command line arguments */
static struct benchmark_clo pmem_flush_clo[] = {
	{
		.opt_short	= 'o',
		.opt_long	= "operation",
		.descr		= "Operation type - persist, msync, ...",
		.type		= CLO_TYPE_STR,
		.off		= clo_field_offset(struct pmem_args, operation),
		.def		= "noop"
	},
	{
		.opt_short	= 0,
		.opt_long	= "mode",
		.descr		= "mode - stat, seq or rand",
		.type		= CLO_TYPE_STR,
		.off		= clo_field_offset(struct pmem_args, mode),
		.def		= "stat",
	},
	{
		.opt_short	= 'w',
		.opt_long	= "no-warmup",
		.descr		= "Don't do warmup",
		.type		= CLO_TYPE_FLAG,
		.off		= clo_field_offset(struct pmem_args, no_warmup),
	},
};

/* Stores information about benchmark. */
static struct benchmark_info pmem_flush_bench = {
	.name		= "pmem_flush",
	.brief		= "Benchmark for pmem_msync() and pmem_persist()",
	.init		= pmem_flush_init,
	.exit		= pmem_flush_exit,
	.multithread	= true,
	.multiops	= true,
	.operation	= pmem_flush_operation,
	.measure_time	= true,
	.clos		= pmem_flush_clo,
	.nclos		= ARRAY_SIZE(pmem_flush_clo),
	.opts_size	= sizeof(struct pmem_args),
	.rm_file	= true,
	.allow_poolset	= false,
};

REGISTER_BENCHMARK(pmem_flush_bench);
